### DISCUSSION TITLE
Booting iPXE kernel from grub2 on UEFI machine

### DISCUSSION THREAD

**Discussion by paulraines68** [ipxe/ipxe#464] [Q&A] [answered]
I am trying to boot iPXE via grub2 on a CentOS8 machine (UEFI) as a way to do remote reinstalls.

I made a grub2 entry that was 

menuentry "Reinstall via iPXE" {
  linuxefi /ipxe.lkrn
  initrdefi /ipxe.script
}

Where ipxe.lkrn and ipxe.script are the kernel and script respectively from the ISO I made which works fine for this purpose when the box is booted from DVD.  And /boot is its on partition and the files exist there.

However, when I get to grub and choose the above entry, I immediately get a General Protection Fault and have to power cycle the box.  Any ideas what is going wrong?

**Comment by NiKiZe** [comment]
`.lkrn` is "pcbios only" linux mode wrapper - as `pcbios` is "pcbios only" that will not work.
On EFI platforms you will need to find a way to run (probably) `ipxe.efi`.

**Reply by NiKiZe** [reply]
To clarify, a efi binary is an efi binary NOT a disk image.
And again, the Linux image format is different (even if there can be compability between the formats, but not in this case)

Here are instructions for running efi binaries from Grub: https://unix.stackexchange.com/questions/277373/is-it-possible-to-use-grub-to-run-an-efi-binary

**Comment by paulraines68** [comment]
Hmm.  Okay but for reasons I don't know I just named it ipxe.lkrn when I copied /mnt/tmp/EFI/BOOT/BOOTX64.EFI off the DVD.  It should be the correct executable and running file shows

```
# file ipxe.lkrn
ipxe.lkrn: MS-DOS executable
```

Should maybe this be used instead?

```
# file /mnt/tmp/images/efiboot.img
/mnt/tmp/images/efiboot.img: x86 boot sector, mkdosfs boot message display, code offset 0x3c, OEM-ID "mkfs.fat", sectors/cluster 4, root entries 512, sectors 20380 (volumes <=32 MB) , Media descriptor 0xf8, sectors/FAT 20, heads 64, serial number 0xcf737bd0, label: "ANACONDA   ", FAT (16 bit)
```

**Reply by NiKiZe** [reply]
Is this in reply to my answer to your question?

**Reply by paulraines68** [reply]
It was but I missed the clarification with the StackExchange link.  Thanks for that.  I got it to work now with this entry:

```
menuentry "Reinstall via iPXE" {
  insmod part_gpt
  insmod chain
  search --no-floppy --fs-uuid --set=root 84d5ce35-af85-44fe-bd15-c26b8418d0d6
  chainloader /ipxe.efi
}
```

**Comment by paulraines68** [comment, marked as answer]
To close this, what works is an entry in /etc/grub.d/40_custom of:

```
menuentry "Reinstall via iPXE" {
  insmod part_gpt
  insmod chain
  search --no-floppy --fs-uuid --set=root 84d5ce35-af85-44fe-bd15-c26b8418d0d6
  chainloader /ipxe.efi
}
```
